consensus/core/src/dag_state.rs

```diff
@@ -970,6 +1081,59 @@ impl DagState {
     }
 }
 
+struct BlockInfo {
+    block: VerifiedBlock,
+    // Accumulates implicit accept votes for all transactions.
+    accept_votes: StakeAggregator<QuorumThreshold>,
+    // Accumulates reject votes per transaction.
+    reject_votes: BTreeMap<TransactionIndex, StakeAggregator<QuorumThreshold>>,
+    // Whether this block has been sent to output after it has been certified.
+    certified_output: bool,
+}
+
+impl BlockInfo {
+    fn new(block: VerifiedBlock) -> Self {
+        Self {
+            block,
+            accept_votes: StakeAggregator::new(),
+            reject_votes: BTreeMap::new(),
+            certified_output: false,
+        }
+    }
+
+    // If this block has been certified but has not been sent to output, returns the output.
+    // Otherwise, returns None.
+    fn take_certified_output(&mut self, committee: &Committee) -> Option<BlockOutput> {
+        if self.certified_output {
+            return None;
+        }
+        if !self.accept_votes.reached_threshold(committee) {
+            return None;
+        }
+        let mut rejected = vec![];
+        for (idx, reject_votes) in &self.reject_votes {
+            if reject_votes.reached_threshold(committee) {
+                rejected.push(*idx);
+                continue;
+            }
+            if self
+                .accept_votes
+                .stake()
+                .checked_sub(reject_votes.stake())
+                .unwrap()
+                < committee.quorum_threshold()
+            {
+                return None;
+            }
```

Comment:
The `rejected_votes.stake()`  refers to the single transaction of index `idx` . Why can't the block be certified in this case? Am I missing something?

Reply:
Yeah I using `certified block` to mean not just the block is certified, but all its transactions are certified to be accepted or rejected. Added comment.

Reply:
I agree `certificate` is already overloaded, for example it can mean certified block in Narwhal, certified transaction in Sui and certified leaders in Mysticeti commit rule.

For a given class of entities (one of blocks, transactions, leaders), do we ever certify them in different ways? I think we do not, so we may just need to be precise on what is being certified. For example in commit rules, we need to use CertifiedLeader / LeaderCertificate instead of just certificate.